parser-typechecker/src/Unison/Codebase/Causal.hs

```diff
@@ -170,17 +213,17 @@ threeWayMerge combine = mergeInternal merge0
  where
   merge0 :: Map (RawHash h) (m (Causal m h e)) -> m (Causal m h e)
   merge0 m =
-    let k left right = do
-          a           <- left
-          b           <- right
-          mayAncestor <- lca a b
-          newHead <- combine (head <$> mayAncestor) (head a) (head b)
-          let h = hash (newHead, Map.keys m)
-          pure . Merge (RawHash h) newHead $ Map.fromList
-            [(currentHash a, pure a), (currentHash b, pure b)]
-    in  if Map.null m
-          then error "Causal.threeWayMerge empty map"
-          else foldl1' k $ Map.elems m
+    let k (e, acc) new = do
+          n           <- new
+          mayAncestor <- lca' acc (Seq.singleton (pure n))
+          newHead     <- combine (head <$> mayAncestor) e (head n)
+          pure (newHead, pure n Seq.<| acc)
+    in  case Map.elems m of
+          []       -> error "Causal.threeWayMerge empty map"
+          me : mes -> do
+            e            <- me
+            (newHead, _) <- foldM k (head e, Seq.singleton (pure e)) mes
+            pure $ Merge (RawHash (hash (newHead, Map.keys m))) newHead m
```

Comment:
For readability (since I'm reading top to bottom, but the context for understanding `k` isn't there yet), would you move `k` below, into a `where clause`, and/or put a type signature on it?